client/packages/system/src/Stock/api/hooks/useStockLine.ts

```diff
@@ -15,76 +16,185 @@ export interface DraftStockLine extends StockLineRowFragment {
   inventoryAdjustmentReason: InventoryAdjustmentReasonRowFragment | null;
 }
 
-export function useStockLine() {
-  const [draft, setDraft] = useState<DraftStockLine>({
-    __typename: 'StockLineNode',
-    id: '',
-    itemId: '',
-    onHold: false,
-    packSize: 0,
-    sellPricePerPack: 0,
-    costPricePerPack: 0,
-    totalNumberOfPacks: 0,
-    availableNumberOfPacks: 0,
-    storeId: '',
-    item: {
-      __typename: 'ItemNode',
-      code: '',
-      name: '',
-    },
-    inventoryAdjustmentReason: null,
-  });
+const defaultDraftStockLine = {
+  __typename: 'StockLineNode',
+  id: '',
+  itemId: '',
+  onHold: false,
+  packSize: 0,
+  sellPricePerPack: 0,
+  costPricePerPack: 0,
+  totalNumberOfPacks: 0,
+  availableNumberOfPacks: 0,
+  storeId: '',
+  item: {
+    __typename: 'ItemNode',
+    code: '',
+    name: '',
+  },
+  inventoryAdjustmentReason: null,
+};
+
+export function useStockLine(id?: string) {
+  const [patch, setPatch] = useState<Partial<DraftStockLine>>({});
+  const [isDirty, setIsDirty] = useState(false);
+  const { data, isLoading, error } = useGet(id ?? '');
+  const {
+    mutateAsync: createMutation,
+    isLoading: isCreating,
+    error: createError,
+  } = useCreate();
+  const {
+    mutateAsync: updateMutation,
+    isLoading: isUpdating,
+    error: updateError,
+  } = useUpdate(id ?? '');
+
+  const draft = (
+    data
+      ? { ...data?.nodes[0], ...patch }
+      : { ...defaultDraftStockLine, ...patch }
+  ) as DraftStockLine;
 
-  const { mutateAsync: createMutation } = useCreate();
+  const updatePatch = (newData: Partial<DraftStockLine>) => {
+    // Only add changed values to patch
+    const changedData = data
+      ? Object.fromEntries(
+          Object.entries(newData).filter(
+            ([key, value]) => value !== draft[key as keyof DraftStockLine]
+          )
+        )
+      : newData;
+    if (Object.keys(changedData).length > 0) {
+      setIsDirty(true);
+      setPatch({ ...patch, ...changedData });
+    }
+  };
+
+  const resetDraft = () => {
+    if (data) {
+      setPatch({});
+      setIsDirty(false);
+    }
+  };
 
-  const create = () => createMutation(draft);
+  const create = () => createMutation(draft).then(() => setIsDirty(false));
+  const update = () => updateMutation(patch).then(() => setIsDirty(false));
 
   return {
+    query: { data: data?.nodes[0], isLoading, error },
+    create: { create, isCreating, createError },
+    update: { update, isUpdating, updateError },
```

Comment:
Don't love this duplicated nested naming -- any better suggestions?

Reply:
Don't really know if it's better, but an idea:
```ts
  data: data?.nodes[0],
  create,
  update,
  loading: { create: loadingCreate, update: loadingUpdate, data: loadingQuery, any: loadingCreate || loadingUpdate || loadingQuery},
  error: { create: createError, update: updateError, data: queryError },
```

Reply:
Seems about the same degree of confusion to me 😆

Might leave it for now, see if the wider group has opinions...